Что выведет программа? Объяснить вывод программы.

```go
package main

type customError struct {
	msg string
}

func (e *customError) Error() string {
	return e.msg
}

func test() *customError {
	{
		// do something
	}
	return nil
}

func main() {
	var err error
	err = test()
	if err != nil {
		println("error")
		return
	}
	println("ok")
}
```

Программа будет выводить "ok".

Это связано с тем, что в Go `nil` для интерфейса и `nil` для конкретного типа - это
не одно и то же. Чтобы интерфейс был `nil`, и его тип, и его значение должны быть `nil`. 
Но в данном случае функция `test` возвращает `nil`, который имеет тип `*customError`,
так что когда он присваивается интерфейсу `error`, сам интерфейс не является `nil`,
несмотря на то, что его значение `nil`.

Однако, в коде переменная `err` объявляется как `error`, и `test()` возвращает
`nil` как `*customError`, который вполне допустимо присваивается `err`. 
Затем проверяется, равно ли `err` `nil`. Поскольку возвращаемое значение `test()`
действительно `nil`, проверка `if err != nil` будет ложной, и поэтому программа 
напечатает "ok".

Таким образом, поведение этого кода отличается от поведения кода в предыдущем вопросе
из-за того, как объявлена переменная `err` и из-за того, что возвращается функцией
`test()`.